File: app/src/main/java/com/partyconsole/companion/network/LiveConnection.kt
package com.partyconsole.companion.network

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.sse.EventSource
import okhttp3.sse.EventSourceListener
import okhttp3.sse.EventSources
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong
import kotlin.coroutines.coroutineContext

/** What a screen actually needs to know about the live connection - either
 *  a character updated (or went offline, record == null), or the
 *  connection's own health changed. Mirrors dashboard-live.tsx's two
 *  observable pieces of state (per-character query data, and the shared
 *  `liveConnectionKey` healthy flag) as one merged event stream, since
 *  Kotlin Flow doesn't need the two separate cache slots React Query used
 *  there for its own reasons. */
sealed interface LiveEvent {
    data class CharacterUpdated(val name: String, val record: LiveRecordWire?) : LiveEvent
    data class ConnectionHealth(val healthy: Boolean) : LiveEvent
}

private const val HEARTBEAT_TIMEOUT_MS = 15_000L
private const val RECONNECT_DELAY_MS = 1_000L
private const val WATCHDOG_TICK_MS = 1_000L

/**
 * Connects to `{baseUrl}/party-api/dashboard-stream` (Server-Sent Events,
 * exactly what the web dashboard's own EventSource connects to - see
 * dashboard-live.tsx) and emits a LiveEvent per update.
 *
 * Two independent triggers can start a reconnect, matching the original's
 * behavior exactly: an immediate one on any stream failure/bad frame
 * (`onFailure`, a JSON parse error), and a 1-second watchdog tick that
 * catches a connection which never errored but also stopped sending
 * heartbeats (a stalled-but-not-closed socket). Both funnel through
 * `scheduleReconnect`, guarded by `reconnectPending` so a failure and a
 * watchdog tick landing close together never schedule two overlapping
 * reconnects - the same single-timer guard the original's `!reconnect`
 * check provides.
 *
 * Cancel the collecting coroutine to close the connection cleanly (e.g.
 * when the app goes to the background) - cleanup runs via awaitClose,
 * same lifecycle guarantee as any other cold Flow.
 */
fun liveEvents(client: OkHttpClient, settings: ServerSettings): Flow<LiveEvent> = callbackFlow {
    val json = Json { ignoreUnknownKeys = true }
    val receiver = LiveReceiver { name, record -> trySend(LiveEvent.CharacterUpdated(name, record)) }
    val lastHeartbeat = AtomicLong(System.currentTimeMillis())
    val reconnectPending = AtomicBoolean(false)
    var currentSource: EventSource? = null
    var stopped = false
    val scope = CoroutineScope(coroutineContext)

    fun reportHealth(healthy: Boolean) {
        trySend(LiveEvent.ConnectionHealth(healthy))
    }

    // startConnection and scheduleReconnect call each other (a failed
    // connection schedules a reconnect, which starts a new connection).
    // Kotlin resolves local `fun` declarations top-to-bottom within a
    // block - unlike class member functions, they can't forward-reference
    // each other, confirmed by an actual compile error here. Declaring
    // both as `var`s of function type first, then assigning both bodies
    // afterward, sidesteps that: each body only needs the OTHER var to
    // already be declared (not yet assigned) at the point it's written,
    // since a lambda body only reads its captured variables when it's
    // actually invoked later, not at declaration time.
    var startConnection: () -> Unit = {}
    var scheduleReconnect: () -> Unit = {}

    // Anonymous `fun()` rather than a `{ }` lambda: a lambda stored in a
    // var doesn't support a bare `return` (that's only valid as a
    // non-local return out of an INLINE function's lambda) - an anonymous
    // function does, so the early-return bodies below work unchanged.
    startConnection = fun() {
        if (stopped) return
        lastHeartbeat.set(System.currentTimeMillis())
        val request = Request.Builder().url(settings.streamUrl).build()
        currentSource = EventSources.createFactory(client).newEventSource(
            request,
            object : EventSourceListener() {
                override fun onOpen(eventSource: EventSource, response: Response) {
                    lastHeartbeat.set(System.currentTimeMillis())
                }

                override fun onEvent(eventSource: EventSource, id: String?, type: String?, data: String) {
                    if (stopped) return
                    val message = runCatching { json.decodeFromString(LiveMessage.serializer(), data) }.getOrNull()
                    if (message == null) {
                        // A single malformed frame is treated the same way
                        // the original's try/catch treats a JSON.parse
                        // failure: close and reconnect, since the epoch/
                        // sequence state can no longer be trusted to be in
                        // sync with the server.
                        eventSource.cancel()
                        reportHealth(false)
                        scheduleReconnect()
                        return
                    }
                    if (!receiver.accept(message)) return
                    // Any accepted message proves the connection is alive -
                    // not just heartbeat/snapshot. The original bug here
                    // only refreshed on those two types, so a character
                    // actively sending delta updates (the normal case
                    // during gameplay) never reset the watchdog; after
                    // HEARTBEAT_TIMEOUT_MS of nothing but real data it
                    // looked "silent" and forced a reconnect anyway.
                    lastHeartbeat.set(System.currentTimeMillis())
                    if (message.type == "snapshot") reportHealth(true)
                }

                override fun onFailure(eventSource: EventSource, t: Throwable?, response: Response?) {
                    if (stopped) return
                    reportHealth(false)
                    scheduleReconnect()
                }
            },
        )
    }

    scheduleReconnect = fun() {
        if (stopped || !reconnectPending.compareAndSet(false, true)) return
        scope.launch {
            delay(RECONNECT_DELAY_MS)
            reconnectPending.set(false)
            if (!stopped) startConnection()
        }
    }

    reportHealth(false)
    startConnection()

    val watchdogJob = scope.launch {
        while (isActive && !stopped) {
            delay(WATCHDOG_TICK_MS)
            val silentFor = System.currentTimeMillis() - lastHeartbeat.get()
            if (silentFor > HEARTBEAT_TIMEOUT_MS) {
                currentSource?.cancel()
                reportHealth(false)
                scheduleReconnect()
            }
        }
    }

    awaitClose {
        stopped = true
        currentSource?.cancel()
        watchdogJob.cancel()
    }
}

File: web/src/api/liveConnection.ts
import { LiveReceiver, type LiveMessage, type LiveRecordWire } from './liveProtocol'
import { streamUrl, type ServerSettings } from '@/config/serverConfig'

/** What a consumer actually needs to know about the live connection -
 *  either a character updated (or went offline, record === null), or the
 *  connection's own health changed. Ports network/LiveConnection.kt's
 *  LiveEvent (itself matching dashboard-live.tsx's two observable pieces
 *  of state: per-character query data, and the shared healthy flag). */
export type LiveEvent =
  | { kind: 'characterUpdated'; name: string; record: LiveRecordWire | null }
  | { kind: 'connectionHealth'; healthy: boolean; error?: string }

const HEARTBEAT_TIMEOUT_MS = 15_000
const RECONNECT_DELAY_MS = 1_000
const WATCHDOG_TICK_MS = 1_000

/**
 * Connects to `{baseUrl}/party-api/dashboard-stream` (Server-Sent Events,
 * exactly what party-console's own web dashboard connects to) and calls
 * [onEvent] per update - a direct port of network/LiveConnection.kt's
 * `liveEvents` Flow, using the browser's native EventSource in place of
 * OkHttp's SSE client.
 *
 * Two independent triggers can start a reconnect, matching the Android
 * port exactly: an immediate one on any stream failure/bad frame, and a
 * 1-second watchdog tick that catches a connection which never errored
 * but also stopped sending heartbeats (a stalled-but-not-closed socket -
 * this is why EventSource's own built-in auto-reconnect isn't relied on
 * alone; it only fires on an actual `error` event, not on silence). Both
 * funnel through `scheduleReconnect`, guarded by `reconnectPending` so a
 * failure and a watchdog tick landing close together never schedule two
 * overlapping reconnects.
 *
 * Call the returned function to close the connection cleanly (e.g. on
 * unmount).
 *
 * Note: the browser's EventSource intentionally exposes NO detail on
 * `onerror` (no HTTP status, no message - a cross-origin/security
 * restriction, unlike OkHttp's onFailure which hands back the real
 * response/exception) - error messages here are necessarily vaguer than
 * the Android app's for connection failures specifically; the malformed-
 * message and heartbeat-timeout paths still carry full detail.
 */
export function connectLive(settings: ServerSettings, onEvent: (event: LiveEvent) => void): () => void {
  const receiver = new LiveReceiver((name, record) => onEvent({ kind: 'characterUpdated', name, record }))
  let lastHeartbeat = Date.now()
  let reconnectPending = false
  let stopped = false
  let currentSource: EventSource | null = null
  let reconnectTimer: ReturnType<typeof setTimeout> | undefined
  let watchdogTimer: ReturnType<typeof setInterval> | undefined

  const reportHealth = (healthy: boolean, error?: string) => onEvent({ kind: 'connectionHealth', healthy, error })

  const scheduleReconnect = () => {
    if (stopped || reconnectPending) return
    reconnectPending = true
    reconnectTimer = setTimeout(() => {
      reconnectPending = false
      if (!stopped) startConnection()
    }, RECONNECT_DELAY_MS)
  }

  const startConnection = () => {
    if (stopped) return
    lastHeartbeat = Date.now()
    const source = new EventSource(streamUrl(settings))
    currentSource = source

    source.onopen = () => {
      lastHeartbeat = Date.now()
    }

    source.onmessage = (event) => {
      if (stopped) return
      let message: LiveMessage | null = null
      try {
        message = JSON.parse(event.data) as LiveMessage
      } catch {
        message = null
      }
      if (message == null) {
        // A single malformed frame: close and reconnect, since the
        // epoch/sequence state can no longer be trusted to be in sync
        // with the server.
        source.close()
        reportHealth(false, 'Received a malformed message from the server')
        scheduleReconnect()
        return
      }
      if (!receiver.accept(message)) return
      // Any accepted message proves the connection is alive - not just
      // heartbeat/snapshot types (a character actively sending delta
      // updates, the normal case during gameplay, must also reset the
      // watchdog, or it looks "silent" and forces a spurious reconnect).
      lastHeartbeat = Date.now()
      if (message.type === 'snapshot') reportHealth(true)
    }

    source.onerror = () => {
      if (stopped) return
      source.close()
      reportHealth(false, describeReadyState(source.readyState))
      scheduleReconnect()
    }
  }

  reportHealth(false)
  startConnection()

  watchdogTimer = setInterval(() => {
    if (stopped) return
    const silentFor = Date.now() - lastHeartbeat
    if (silentFor > HEARTBEAT_TIMEOUT_MS) {
      currentSource?.close()
      reportHealth(false, `No response from the server for over ${HEARTBEAT_TIMEOUT_MS / 1000}s`)
      scheduleReconnect()
    }
  }, WATCHDOG_TICK_MS)

  return () => {
    stopped = true
    currentSource?.close()
    if (reconnectTimer) clearTimeout(reconnectTimer)
    if (watchdogTimer) clearInterval(watchdogTimer)
  }
}

function describeReadyState(state: number): string {
  switch (state) {
    case EventSource.CONNECTING:
      return 'Connection failed while connecting (check the address and that the server is reachable)'
    case EventSource.CLOSED:
      return 'Connection closed unexpectedly'
    default:
      return 'Connection failed for an unknown reason'
  }
}
